server/src/sim.rs

```diff
@@ -242,42 +254,69 @@ impl Sim {
     fn snapshot_node(&self, node: NodeId) -> save::EntityNode {
         let mut entities = Vec::new();
         for &entity in self.graph_entities.get(node) {
-            let Ok(entity) = self.world.entity(entity) else {
-                error!("stale graph entity {:?}", entity);
-                continue;
-            };
-            let Some(id) = entity.get::<&EntityId>() else {
-                continue;
-            };
-            let mut components = Vec::new();
-            if let Some(pos) = entity.get::<&Position>() {
-                components.push((
-                    ComponentType::Position as u64,
-                    postcard::to_stdvec(&pos.local.as_ref()).unwrap(),
-                ));
-            }
-            if let Some(ch) = entity.get::<&Character>().or_else(|| {
-                entity
-                    .get::<&InactiveCharacter>()
-                    .map(|ich| hecs::Ref::map(ich, |ich| &ich.0)) // Extract Ref<Character> from Ref<InactiveCharacter>
-            }) {
-                components.push((ComponentType::Name as u64, ch.name.as_bytes().into()));
-            }
-            let mut repr = Vec::new();
-            postcard_helpers::serialize(
-                &SaveEntity {
-                    entity: id.to_bits().to_le_bytes(),
-                    components,
-                },
-                &mut repr,
-            )
-            .unwrap();
-            entities.push(repr);
+            let reprs = self.snapshot_entity_and_children(entity);
+            entities.extend_from_slice(&reprs);
         }
 
         save::EntityNode { entities }
     }
 
+    fn snapshot_entity_and_children(&self, entity: Entity) -> Vec<Vec<u8>> {
```

Comment:
Factor out the function in one commit, then modify its behavior in a separate commit. When those two changes are in the same commit it's much harder to identify behavior changes.

Reply:
Good point. Depending on how https://github.com/Ralith/hypermine/pull/431#discussion_r1797768956 is resolved, I may end up undoing the refactor.

Reply:
I have now undone this refactor.